.drone.starlark

```diff
@@ -0,0 +1,950 @@
+			extraSetup = matrix['extraSetup'] if 'extraSetup' in matrix else default['extraSetup']
+			extraServices = matrix['extraServices'] if 'extraServices' in matrix else default['extraServices']
+			extraApps = matrix['extraApps'] if 'extraApps' in matrix else default['extraApps']
+
+			for server in servers:
+				for browser in browsers:
+					for phpVersion in phpVersions:
+						for db in databases:
+							name = 'unknown'
+
+							if isWebUI or isAPI or isCLI:
+								browserString = '' if browser == '' else '-' + browser
+								name = '%s-%s%s-%s-php%s' % (shortSuiteName, server.replace('daily-', '').replace('-qa', ''), browserString, db.replace(':', ''), phpVersion)
+								maxLength = 50
+								nameLength = len(name)
+								if nameLength > maxLength:
+									print("Error: generated stage name of length", nameLength, "is not supported. The maximum length is " + str(maxLength) + ".", name)
+									errorFound = True
+
+							if isWebUI:
+								environment = {
+									'TEST_SERVER_URL': 'http://server',
+									'BEHAT_FILTER_TAGS': filterTags,
+									'BEHAT_SUITE': suite,
+									'SELENIUM_HOST': 'selenium',
+									'SELENIUM_PORT': '4444',
+									'BROWSER': browser,
+									'PLATFORM': 'Linux',
+								}
+
+							if isAPI or isCLI:
+								environment = {
+									'TEST_SERVER_URL': 'http://server',
+									'BEHAT_FILTER_TAGS': filterTags,
+									'BEHAT_SUITE': suite,
+								}
+
+							if emailNeeded:
+								environment['MAILHOG_HOST'] = 'email'
+
+							result = {
+								'kind': 'pipeline',
+								'type': 'docker',
+								'name': name,
+								'workspace' : {
+									'base': '/var/www/owncloud',
+									'path': 'server/apps/%s' % config['app']
+								},
+								'steps': [
+									installCore(server, db),
+									installTestrunner(phpVersion)
+								] + ([
+									{
+										'name': 'install-federation',
+										'image': 'owncloudci/core',
+										'pull': 'always',
+										'settings': {
+											'version': server,
+											'core_path': '/var/www/owncloud/federated'
+										}
+									},
+									{
+										'name': 'configure-federation',
+										'image': 'owncloudci/php:%s' % phpVersion,
+										'pull': 'always',
+										'commands': [
+											'echo "export TEST_SERVER_FED_URL=http://federated" > /var/www/owncloud/saved-settings.sh',
+											'cd /var/www/owncloud/federated',
+											'php occ a:l',
+											'php occ a:e testing',
+											'php occ a:l',
+											'php occ config:system:set trusted_domains 1 --value=federated',
+											'php occ log:manage --level 0',
+											'php occ config:list'
+										]
+									}
+								] if federatedServerNeeded else []) + [
+									setupServerAndApp(phpVersion),
+									installExtraApps(phpVersion, extraApps),
+									extraSetup,
+									fixPermissions(phpVersion),
+									({
+										'name': 'acceptance-tests',
```

Comment:
can we join this blocks together for API, WebUI & CLI tests and only have a different `make` command for each?

Reply:
done